integration-tests/contract-staking.spec.ts

```diff
@@ -0,0 +1,102 @@
+import { Protocols } from "@taquito/taquito";
+import { CONFIGS, isSandbox, sleep } from "./config";
+
+CONFIGS().forEach(({ lib, rpc, setup, protocol }) => {
+    const Tezos = lib;
+    describe(`Test staking through contract API using: ${rpc}`, () => {
+      const flextesaOxford = isSandbox({rpc}) && protocol === Protocols.ProxfordS ? test : test.skip;
+      beforeEach(async (done) => {
+          await setup(true)
+          done()
+        });
+        flextesaOxford('Should be able to stake', async (done) => {
+          const op = await Tezos.contract.stake({
+            amount: 100000,
+          });
+          await op.confirmation()
+          expect(op.hash).toBeDefined();
+          
+          const constants = await Tezos.rpc.getConstants();
+          await sleep(((constants.preserved_cycles + 2) * constants.blocks_per_cycle * (constants.minimal_block_delay!.toNumber())) * 1000);
+          
+          done();
+        });
+        flextesaOxford('Should be able to unstake', async (done) => {
+          const op = await Tezos.contract.unstake({
+            amount: 100000,
+          });
+          await op.confirmation()
+          expect(op.hash).toBeDefined();
+
+          const constants = await Tezos.rpc.getConstants();
+          await sleep(((constants.preserved_cycles + 2) * constants.blocks_per_cycle * (constants.minimal_block_delay!.toNumber())) * 1000);
+
+          done();
+        });
+        flextesaOxford('Should be able to finalizeUnstake', async (done) => {
+          const op = await Tezos.contract.finalizeUnstake({});
+          await op.confirmation()
+          expect(op.hash).toBeDefined();
+          done();
+        });
+        it.skip('Should fail with unstake amount too large, could not produce: gets proto.018-Proxford.tez.subtraction_underflow when the amount is too high', async (done) => {
+          expect(async () => {
+            const op = await Tezos.contract.stake({
+              amount: 1000000000000,
+            });
+            await op.confirmation()
+          }).rejects.toThrowError(/Invalid unstake request amount/);
+          done();
+        });
+        flextesaOxford('Should fail with invalid_nonzero_transaction_amount', async (done) => {
+          expect(async () => {
+            const address = await Tezos.signer.publicKeyHash();
+            const op = await Tezos.contract.transfer({
+              to: address,
+              amount: 10000,
```

Comment:
if you use contract transfer to construct an unstake op, amount should be 0

Reply:
This is the intention. We are testing that we can see this error.